cartographer/mapping/sparse_pose_graph.cc

```diff
@@ -74,32 +80,52 @@ proto::SparsePoseGraphOptions CreateSparsePoseGraphOptions(
 
 proto::SparsePoseGraph SparsePoseGraph::ToProto() {
   proto::SparsePoseGraph proto;
+
+  std::vector<std::vector<TrajectoryNode>> grouped_nodes;
+  std::vector<std::pair<int, int>> new_indices;
+  GroupTrajectoryNodes(GetTrajectoryNodes(), &grouped_nodes, &new_indices);
+
   for (const auto& constraint : constraints()) {
     auto* const constraint_proto = proto.add_constraint();
     *constraint_proto->mutable_relative_pose() =
         transform::ToProto(constraint.pose.zbar_ij);
+    constraint_proto->mutable_sqrt_lambda()->Reserve(36);
     for (int i = 0; i != 36; ++i) {
       constraint_proto->mutable_sqrt_lambda()->Add(0.);
     }
     Eigen::Map<Eigen::Matrix<double, 6, 6>>(
         constraint_proto->mutable_sqrt_lambda()->mutable_data()) =
         constraint.pose.sqrt_Lambda_ij;
-    // TODO(whess): Support multi-trajectory.
-    constraint_proto->mutable_submap_id()->set_submap_index(constraint.i);
-    constraint_proto->mutable_scan_id()->set_scan_index(constraint.j);
+
+    constraint_proto->mutable_submap_id()->set_trajectory_id(
+        new_indices[constraint.i].first);
```

Comment:
If I understand correctly, my `j` values (the scan indices) are correct; each `TrajectoryNode` owns a single scan, and therefore the `new_indices` mapping gets us there. Right?

The submap indices are quite a bit trickier, because each `TrajectoryNode` (and therefore each trajectory_id) has a `Submaps*`, but doesn't know which `Submap` it belongs to.

I think. 

Right?